### PR TITLE
Operations for networks app such as search, color and labelNodes

### PR DESCRIPTION
**Pull request by senakafdo**
This functionality is required for making it possible to search within network diagrams using Sigma.js.

### CODE REVIEW

File: packages/ove-app-networks/src/swagger-extensions.yaml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+      parameters:
+      - name: "filter"
+        in: "query"
+        type: "string"
+        description: "Filter for nodes"
+        required: false
+      - name: "color"
+        in: "query"
+        type: "string"
+        description: "New color for nodes in the format rgb(x, y, z)"
```
Comment (davidbirchwork) [inline packages/ove-app-networks/src/swagger-extensions.yaml]: An example here would help, why x,y,z rather than r,g,b? int or hex code accepted?

File: packages/ove-app-networks/src/client/constants/networks.js [hunk 2]
```diff
@@ -4,11 +4,63 @@ const Constants = {
...
+    SEARCH: 'search',
+    COLOR: 'color',
+    LABEL: 'labelNodes',
+    NEIGHBORS_OF: 'neighborsOf'
+};
+
+Constants.Evaluation = {
+    PROPERTY: 'property',
+    LITERAL: 'literal',
+    FUNCTION_CALL: 'functionCall'.toLowerCase(),
```
Comment (davidbirchwork) [inline packages/ove-app-networks/src/client/constants/networks.js]: can this be simplified?

File: packages/ove-app-networks/src/client/constants/networks.js [hunk 3]
```diff
@@ -4,11 +4,63 @@ const Constants = {
...
+    ADD: 'add',
+    SUBTRACT: 'sub',
+    MULTIPLY: 'mul',
+    DIVIDE: 'div',
+    MODULO: 'mod',
+
+    // The OData specification requires lower-case function names.
+    Function: {
+        SUBSTRING: 'substring',
+        SUBSTRING_OF: 'substringOf'.toLowerCase(),
```
Comment (davidbirchwork) [inline packages/ove-app-networks/src/client/constants/networks.js]: why not remove the toLowerCase() call in order to improve performance?

File: packages/ove-app-networks/src/client/common/networks.js [hunk 4]
```diff
@@ -13,6 +13,235 @@ $(function () {
...
+                const nodeLabel = getFromMessage(message, 'node', 'label');
+                log.debug('Applying label:', nodeLabel, 'on all nodes');
+                f = f.nodesBy(function (n) {
+                    n.label = getFromElement(n, nodeLabel);
+                    return true;
+                });
+            }
+            f.apply();
+            break;
+        case Constants.Operation.NEIGHBORS_OF:
```
Comment (davidbirchwork) [inline packages/ove-app-networks/src/client/common/networks.js]: https://grammarist.com/spelling/neighbor-neighbour/ 
may we not be American please ;-)

File: packages/ove-app-networks/src/client/common/networks.js [hunk 5]
```diff
@@ -13,6 +13,235 @@ $(function () {
...
+        const evaluateN = function (element, filter) {
+            return +(evaluate(element, filter));
+        };
+        // Evaluate as a string
+        const evaluateS = function (element, filter) {
+            return evaluate(element, filter).toString();
+        };
+        // Evaluate a function
+        const evaluateF = function (element, func, args) {
+            switch (func) {
```
Comment (davidbirchwork) [inline packages/ove-app-networks/src/client/common/networks.js]: could the switch be reordered to group by #operands? then argument length checking may be done per batch?

File: packages/ove-app-networks/src/client/common/networks.js [hunk 6]
```diff
@@ -13,6 +13,235 @@ $(function () {
...
+    // IMPORTANT: There are no logs within the filter evaluation operations to ensure
+    // the most optimum performance. The logs corresponding to the operation carried out
+    // can be used for debugging purposes.
+    const evaluate = function (element, filter) {
+        // Evaluate as a number
+        const evaluateN = function (element, filter) {
+            return +(evaluate(element, filter));
+        };
+        // Evaluate as a string
+        const evaluateS = function (element, filter) {
```
Comment (davidbirchwork) [inline packages/ove-app-networks/src/client/common/networks.js]: is there performance value in lifting these function definitions out (above) this scope?

File: packages/ove-app-networks/src/client/common/networks.js [hunk 7]
```diff
@@ -13,6 +13,235 @@ $(function () {
...
+        let m;
+        let n;
+        if (!x || x.indexOf('.') === -1) {
+            n = x;
+        } else {
+            m = x.substring(x.indexOf('.') + 1);
+            n = x.substring(0, x.indexOf('.'));
+        }
+        const p = e[n] || (e.attributes ? e.attributes[n] : undefined);
+        return m ? getFromElement(p, m) : p;
```
Comment (jamesscottbrown) [inline packages/ove-app-networks/src/client/common/networks.js]: I think the logic would be clearer if the ternary was replaced by moving the logic into if statement:

```JavaScript
        let m, n;
        if (!x || x.indexOf('.') === -1) {
            return e[x] || (e.attributes ? e.attributes[x] : undefined);
        } else {
            m = x.substring(x.indexOf('.') + 1);
            n = x.substring(0, x.indexOf('.'));
            const p = e[n] || (e.attributes ? e.attributes[n] : undefined);
            return getFromElement(p, m);
        } 
```

File: packages/ove-app-networks/src/client/common/networks.js (hunk 4 above)
Comment (senakafdo) [reply packages/ove-app-networks/src/client/common/networks.js]: @davidbirchwork, please have a look at https://github.com/jacomyal/sigma.js/tree/master/plugins/sigma.plugins.filter. We may not be American, but we cannot have two spellings for the same operation.

File: packages/ove-app-networks/src/client/common/networks.js (hunk 5 above)
Comment (senakafdo) [reply packages/ove-app-networks/src/client/common/networks.js]: That unwantedly complicates the entire implementation and makes it very hard to follow. These are all string operations, and therefore no further grouping would be required. The only reason why we have argument length checking for the substring operation is because the spec allows two variants of it.

File: packages/ove-app-networks/src/client/common/networks.js (hunk 6 above)
Comment (senakafdo) [reply packages/ove-app-networks/src/client/common/networks.js]: No, they would work out be the same when compiled. You can try checking by generating a min out of this JS file. The format that we have right now makes it easy to read and follow.

File: packages/ove-app-networks/src/client/constants/networks.js (hunk 2 above)
Comment (senakafdo) [reply packages/ove-app-networks/src/client/constants/networks.js]: I'm assuming the simplification is to get rid of the `toLowerCase` conversion, which is also questioned below. I have addressed that there, and therefore resolving this comment.

File: packages/ove-app-networks/src/client/common/networks.js [hunk 8]
```diff
@@ -13,6 +13,235 @@ $(function () {
...
+                const err = 'Unable to evaluate unknown type: ' + filter.type;
+                log.error(err);
+                throw Error(err);
+        }
+    };
+
+    (function () {
+        // This function resets colors and labels to the original state before
+        // running any further filtering. Edges don't have labels by default and
+        // reducing the parsing overhead can improve performance for some graphs.
```
Comment (jamesscottbrown) [inline packages/ove-app-networks/src/client/common/networks.js]: What you mean by 'parsing overhead' here? The parsing has already occurred (on the server-side).

File: packages/ove-app-networks/src/swagger-extensions.yaml [hunk 9]
```diff
@@ -0,0 +1,120 @@
+tags:
+- name: "operation"
+  description: "Operations that can be performed on one or all networks"
+paths:
+  /operation/search:
+    get:
+      tags:
+      - "operation"
+      summary: "Searches for nodes or edges that match provider filters. Filters must be specified according to the OData v3.0 $filter specification."
```
Comment (jamesscottbrown) [inline packages/ove-app-networks/src/swagger-extensions.yaml]: I think this is a bit misleading: it doesn't just search, it *filters* the display to hide nodes or edges that do not match a filter.

File: packages/ove-app-networks/src/swagger-extensions.yaml (hunk 1 above)
Comment (senakafdo) [reply packages/ove-app-networks/src/swagger-extensions.yaml]: I believe that this can be improved to match https://github.com/jacomyal/sigma.js/blob/4ce817bd6f66f24f56a94965afdbebe5226ef9f6/src/utils/sigma.utils.js#L116, which is ulimately what we support. However, the hash cannot be used in a URL, and therefore, we will have to explicitly add that as the first character.

File: packages/ove-app-networks/src/client/common/networks.js [hunk 10]
```diff
@@ -13,6 +13,235 @@ $(function () {
...
+            break;
+        case Constants.Operation.COLOR:
+            if (nodeFilter) {
+                const nodeColor = getFromMessage(message, 'node', 'color');
+                log.debug('Applying color:', nodeColor, 'on all nodes matching filter:', nodeFilter);
+                f = f.nodesBy(function (n) {
+                    if (evaluate(n, nodeFilter)) {
+                        n.color = nodeColor;
+                    }
+                    return true;
```
Comment (jamesscottbrown) [inline packages/ove-app-networks/src/client/common/networks.js]: Perhaps add a comment here:

```JavaScript
    return true; // select all nodes, so none are hidden
```
and similarly for the other equivalent statements

File: packages/ove-app-networks/src/client/constants/networks.js (hunk 3 above)
Comment (senakafdo) [reply packages/ove-app-networks/src/client/constants/networks.js]: There is really no performance difference since these are constants. Not having the function called explicitly did make linters and spellcheckers call these out and potentially introduce a mistaken fix (could even be automatic in some tools). The spec is very strict on the case of these names, and therefore it will instantly break the logic. This is why I included the `toLowerCase` function that makes the code much more resilient and readable.

File: packages/ove-app-networks/README.md [hunk 11]
```diff
@@ -41,3 +41,13 @@ The networks app has a transparent background. If required, a background colour
 ## Controlling the App
 
 The controller of the app can be loaded by accessing the URL `http://OVE_APP_NETWORKS_HOST:PORT/control.html?oveSectionId=SECTION_ID`.
+
```
Comment (jamesscottbrown) [inline packages/ove-app-networks/README.md]: We should state in the documentation that filters are chained (because we use statements like `f = f.nodesBy(...)`), and this has the effect of combining applied filters with a logical AND.

There should also be an API call to clear all filters.

File: packages/ove-app-networks/README.md [hunk 12]
```diff
@@ -41,3 +41,13 @@ The networks app has a transparent background. If required, a background colour
...
+The app also exposes controls as a part of its [API](http://OVE_APP_NETWORKS_HOST:PORT/api-docs#operation). The API exposes operations such as `search`, `color`, `labelNodes`, `neighborsOf`. These operations can be executed on a per-network basis or across all networks. The filters used for nodes and edges confirm to the [OData v3.0 specification](https://www.odata.org/documentation/odata-version-3-0/odata-version-3-0-core-protocol/#thefiltersystemqueryoption) and the colors used must have the format `rgb(x, y, z)`.
```
Comment (jamesscottbrown) [inline packages/ove-app-networks/README.md]: The app **'s**  ~~also exposes controls as a part of its~~ [API](http://OVE_APP_NETWORKS_HOST:PORT/api-docs#operation)~~. The API~~ **also** exposes operations such as `search`, `color`, `labelNodes`, `neighborsOf`.

Comment (jamesscottbrown) [inline packages/ove-app-networks/README.md]: "`rgb(x, y, z)` **(where `x, y, z` are integers in the range 0-255)**"

(a common alternative representation is floats in the range 0 to 1)

**Comment by senakafdo** [comment]
@davidbirchwork, queries are never a part of the state of anything that we have implemented in any framework so far, I don't see the reason for making an exception here. It complicates the logic and may require introducing unwanted timeouts etc, because there is absolutely no way where our JS code can figure out whether Sigma has loaded the graphs or not.

However, this does not limit any visualisations that we build that would use these queries. For example, if we are exploring Tweets it is more than sensible to have specific queries stored and recorded to avoid the user having to type in their search terms over and over. This however is not something that can be (or even needs to be) generically implemented in the OVE app. It is important to understand this distinction.

File: packages/ove-app-networks/README.md [hunk 13]
```diff
@@ -41,3 +41,13 @@ The networks app has a transparent background. If required, a background colour
...
+
+To search for nodes having a size greater than or equal to 2, using OVE APIs:
+
+```sh
+curl  --request GET http://OVE_APP_NETWORKS_HOST:PORT/operation/search?filter=size%20ge%202
```
Comment (jamesscottbrown) [inline packages/ove-app-networks/README.md]: I think it might be more appropriate for these API requests to use the ``POST`` method rather``GET`` method

File: packages/ove-app-networks/src/client/common/networks.js (hunk 5 above)
Comment (jamesscottbrown) [reply packages/ove-app-networks/src/client/common/networks.js]: I don't think we need to re-order the switch statement, but I think there is a benefit to having some logic based on the arity of the function.

The main thing making the current code hard to read is the frequent repetition of `evaluateS(element, args[0])` everywhere, with the occasional `evaluateS(element, args[1])` and `evaluateS(element, args[2])`.

This could be avoided by something like:

```JavaScript
const firstArg = evaluateS(element, args[0]);

if (args.length > 1){
   const secondArg = evaluateS(element, args[1]);
}

if (args.length > 2){
   const thirdArg = evaluateS(element, args[2]);
}
```

Comment (davidbirchwork) [reply packages/ove-app-networks/src/client/common/networks.js]: @jamesscottbrown good plan, that solves both concerns

**Comment by senakafdo** [comment]
> @davidbirchwork, queries are never a part of the state of anything that we have implemented in any framework so far, I don't see the reason for making an exception here. It complicates the logic and may require introducing unwanted timeouts etc, because there is absolutely no way where our JS code can figure out whether Sigma has loaded the graphs or not.

@jamesscottbrown and I did some research on this, and it seems that Sigma parsers do provide a callback function, which we infact are using in our code to refresh/load the graph once parsing is complete. We can run our queries at this point, without having to put in place any timeouts. Given that there are no animations etc, we do not need to ensure we have loaded the graphs across all the screens at the point we run the queries - which makes the implementation straightforward.

File: packages/ove-app-networks/src/index.js [hunk 14]
```diff
@@ -1,11 +1,116 @@
...
+        ws.on('close', function () {
+            log.warn('Lost websocket connection attempting to reconnect');
+            // If the socket is closed, we try to refresh it.
+            setTimeout(getSocket, Constants.SOCKET_REFRESH_DELAY);
+        });
+    };
+    getSocket();
+}, Constants.SOCKET_READY_WAIT_TIME);
+
+let operationsList = Constants.Operation.SEARCH + '|' + Constants.Operation.COLOR + '|' +
```
[the remark this answers is not in the document]
Comment (senakafdo) [reply packages/ove-app-networks/src/index.js]: Can use the following: 
```JavaScript
let operationsList = Object.values(Constants.Operation).join('|');
```

File: packages/ove-app-networks/README.md (hunk 13 above)
Comment (senakafdo) [reply packages/ove-app-networks/README.md]: This is a good suggestion, but for the sake of consistency, all other apps should also be using `POST` instead of `GET` methods, which is a breaking change. This cannot be done in this pull request.

File: packages/ove-app-networks/src/swagger-extensions.yaml (hunk 1 above)
Comment (senakafdo) [reply packages/ove-app-networks/src/swagger-extensions.yaml]: James has suggested a fix to this and I have incorporated that in my fixes.

File: packages/ove-app-networks/src/client/common/networks.js (hunk 8 above)
Comment (senakafdo) [reply packages/ove-app-networks/src/client/common/networks.js]: That sentence is wrong. I think I had something here previously and moved it elsewhere but did not fix it properly afterwards. I have gotten rid of this line altogether and re-written it properly.

**Comment by davidbirchwork** [comment]
Thank you @senakafdo this looks great, the ability to save the state of the app including the current query is a really useful feature for our users, thanks for including this.
